Sources/Subprocess/IO/AsyncIO+Windows.swift

```diff
@@ -0,0 +1,437 @@
+                    handle, ioPort, completionKey, 0
+                ) == ioPort else {
+                    let error = SubprocessError(
+                        code: .init(.asyncIOFailed("CreateIoCompletionPort failed")),
+                        underlyingError: .init(rawValue: GetLastError())
+                    )
+                    continuation.finish(throwing: error)
+                    return
+                }
+                // Now save the continuation
+                _registration.withLock { storage in
+                    storage[completionKey] = continuation
+                }
+            case .failure(let error):
+                continuation.finish(throwing: error)
+            }
+        }
+    }
+
+    internal func removeRegistration(for handle: HANDLE) {
+        let completionKey = UInt64(UInt(bitPattern: handle))
+        _registration.withLock { storage in
+            storage.removeValue(forKey: completionKey)
+        }
+    }
+
+    func read(
+        from diskIO: borrowing IOChannel,
+        upTo maxLength: Int
+    ) async throws -> [UInt8]? {
+        return try await self.read(from: diskIO.channel, upTo: maxLength)
+    }
+
+    func read(
+        from handle: HANDLE,
+        upTo maxLength: Int
+    ) async throws -> [UInt8]? {
+        // If we are reading until EOF, start with readBufferSize
+        // and gradually increase buffer size
+        let bufferLength = maxLength == .max ? readBufferSize : maxLength
+
+        var resultBuffer: [UInt8] = Array(
+            repeating: 0, count: bufferLength
+        )
+        var readLength: Int = 0
+        var signalStream = self.registerHandle(handle).makeAsyncIterator()
+
+        while true {
+            var overlapped = _OVERLAPPED()
+            let succeed = try resultBuffer.withUnsafeMutableBufferPointer { bufferPointer in
+                // Get a pointer to the memory at the specified offset
+                // Windows ReadFile uses DWORD for target count, which means we can only
+                // read up to DWORD (aka UInt32) max.
+                let targetCount: DWORD
+                if MemoryLayout<Int>.size == MemoryLayout<Int32>.size {
+                    // On 32 bit systems we don't have to worry about overflowing
+                    targetCount = DWORD(truncatingIfNeeded: bufferPointer.count - readLength)
+                } else {
+                    // On 64 bit systems we need to cap the count at DWORD max
+                    targetCount = DWORD(truncatingIfNeeded: min(bufferPointer.count - readLength, Int(UInt32.max)))
+                }
+
+                let offsetAddress = bufferPointer.baseAddress!.advanced(by: readLength)
+                // Read directly into the buffer at the offset
+                return ReadFile(
+                    handle,
+                    offsetAddress,
+                    DWORD(truncatingIfNeeded: targetCount),
+                    nil,
+                    &overlapped
+                )
```

Comment:
The `OVERLAPPED` struct appears to be all zero at this point. This is fine for non-seekable handles, like pipes, but for seekable handles (like file handles), this code is wrong as in that case you need to set the offset at which the operation should take place (Win32 doesn't keep track of the file pointer for you for overlapped I/O).

Given the context, I'm pretty sure we're using it with pipes, so this isn't a bug per-se, but we should be cautious about reusing it elsewhere. It would be good to have a comment explaining this above the `ReadFile` and `WriteFile` calls in this file.

Reply:
Yes it's only used with pipes (non-seekable). I'll add a comment.